tasty-query/shared/src/main/scala/tastyquery/reader/classfiles/ClassfileParser.scala

```diff
@@ -324,6 +333,88 @@ private[reader] object ClassfileParser {
+  end readAnnotations
+
+  private def readAnnotations(
+    sym: TermOrTypeSymbol,
+    annotationsStream: Option[Forked[DataStream]]
+  )(using ConstantPool, ReaderContext, InnerClasses, Resolver): List[Annotation] =
+    annotationsStream.fold(Nil)(readAnnotations(sym, _))
+  end readAnnotations
+
+  private def readAnnotations(
+    sym: TermOrTypeSymbol,
+    annotationsStream: Forked[DataStream]
+  )(using ConstantPool, ReaderContext, InnerClasses, Resolver): List[Annotation] =
+    val classfileAnnots = annotationsStream.use(ClassfileReader.readAllAnnotations())
+    classfileAnnots.map(classfileAnnotToAnnot(_))
+
+  private def classfileAnnotToAnnot(
+    classfileAnnot: CFAnnotation
+  )(using ConstantPool, ReaderContext, InnerClasses, Resolver): Annotation =
+    val annotationType = JavaSignatures.parseFieldDescriptor(classfileAnnot.tpe.name)
+
+    val args: List[TermTree] =
+      for (name, value) <- classfileAnnot.values.toList yield
+        val valueTree = annotationValueToTree(value)
+        NamedArg(name, valueTree)(SourcePosition.NoPosition)
+
+    Annotation.fromAnnotTypeAndArgs(annotationType, args)
+  end classfileAnnotToAnnot
+
+  private def annotationValueToTree(
+    value: AnnotationValue
+  )(using ConstantPool, ReaderContext, InnerClasses, Resolver): TermTree =
+    import AnnotationValue.Tags
+
+    val pool = summon[ConstantPool]
+    val pos = SourcePosition.NoPosition
+
+    value match
+      case AnnotationValue.Const(tag, valueIdx) =>
+        val constant = (tag: @switch) match
+          case Tags.Byte    => Constant(pool.integer(valueIdx).toByte)
+          case Tags.Char    => Constant(pool.integer(valueIdx).toChar)
+          case Tags.Double  => Constant(pool.double(valueIdx))
+          case Tags.Float   => Constant(pool.float(valueIdx))
+          case Tags.Int     => Constant(pool.integer(valueIdx))
+          case Tags.Long    => Constant(pool.long(valueIdx))
+          case Tags.Short   => Constant(pool.integer(valueIdx).toShort)
+          case Tags.Boolean => Constant(pool.integer(valueIdx) != 0)
+          case Tags.String  => Constant(pool.utf8(valueIdx).name)
+        Literal(constant)(pos)
+
+      case AnnotationValue.EnumConst(descriptor, constName) =>
+        /* JVMS says that it can be any field descriptor,
+         * but I don't see what we would do with a base type or array type.
+         */
+        val binaryName = descriptor.name match
+          case s"L$binaryName;" => binaryName
+          case other            => throw ClassfileFormatException(s"unexpected non-class field descriptor: $other")
+        val enumClassStaticRef = resolver.resolveStatic(termName(binaryName))
+        val constRef = TermRef(enumClassStaticRef, constName)
+        Ident(constName)(constRef)(pos)
+
+      case AnnotationValue.ClassConst(descriptor) =>
+        val classType = JavaSignatures.parseReturnDescriptor(descriptor.name)
+        Literal(Constant(classType))(pos)
```

Comment:
documentation of Constant should mention that this can be a raw class type

Reply:
maybe add a test for a generic external class

Reply:
Added documentation + test.